Realm/RLMLogger.mm

```diff
@@ -0,0 +1,192 @@
+        case RLMLogLevelDebug:  return Level::debug;
+        case RLMLogLevelTrace:  return Level::trace;
+        case RLMLogLevelAll:    return Level::all;
+    }
+    REALM_UNREACHABLE();    // Unrecognized log level.
+}
+
+static RLMLogLevel logLevelForLevel(Level logLevel) {
+    switch (logLevel) {
+        case Level::off:    return RLMLogLevelOff;
+        case Level::fatal:  return RLMLogLevelFatal;
+        case Level::error:  return RLMLogLevelError;
+        case Level::warn:   return RLMLogLevelWarn;
+        case Level::info:   return RLMLogLevelInfo;
+        case Level::detail: return RLMLogLevelDetail;
+        case Level::debug:  return RLMLogLevelDebug;
+        case Level::trace:  return RLMLogLevelTrace;
+        case Level::all:    return RLMLogLevelAll;
+    }
+    REALM_UNREACHABLE();    // Unrecognized log level.
+}
+
+static NSString* levelPrefix(Level logLevel) {
+    switch (logLevel) {
+        case Level::off:
+        case Level::all:    return @"";
+        case Level::trace:  return @"Trace";
+        case Level::debug:  return @"Debug";
+        case Level::detail: return @"Detail";
+        case Level::info:   return @"Info";
+        case Level::error:  return @"Error";
+        case Level::warn:   return @"Warning";
+        case Level::fatal:  return @"Fatal";
+    }
+    REALM_UNREACHABLE();    // Unrecognized log level.
+}
+
+struct CocoaLogger : public Logger {
+    void do_log(Level level, const std::string& message) override {
+        NSLog(@"%@: %@", levelPrefix(level), RLMStringDataToNSString(message));
+    }
+};
+
+class CustomLogger : public Logger {
+public:
+    RLMLoggerFunction function;
+    void do_log(Level level, const std::string& message) override {
+        @autoreleasepool {
+            if (function) {
+                function(logLevelForLevel(level), RLMStringDataToNSString(message));
+            }
+        }
+    }
+};
+
+struct LoggerFunctionCallbackWrapper {
+    void (^block)(RLMLogLevel, NSString *);
+};
+
+void RLMSetDefaultLogLevel(Level level) {
+    Logger::set_default_level_threshold(level);
+}
+} // anonymous namespace
+
+@implementation RLMLogger {
+    std::shared_ptr<Logger> _logger;
+}
+
+static RLMLogger *_defaultLogger;
+static RLMLogLevel _defaultLogLevel = RLMLogLevelWarn;
+
+- (RLMLogLevel)level {
+    return logLevelForLevel(_logger->get_level_threshold());
+}
+
+- (void)setLevel:(RLMLogLevel)logLevel {
+    _logger->set_level_threshold(levelForLogLevel(logLevel));
+}
+
+- (instancetype)initWithLogger:(std::shared_ptr<Logger>)logger {
+    if (self = [self init]) {
+        self->_logger = logger;
+    }
+    return self;
+}
+
+- (instancetype)init {
+    if (self = [super init]) {
+        auto logger = std::make_shared<CustomLogger>();
+        auto wrapper = LoggerFunctionCallbackWrapper{(^(RLMLogLevel level, NSString *message) {
+            [self doLog:level message:message];
+        })};
+        logger->function = wrapper.block;
```

Comment:
Why is this constructing a LoggerFunctionCallbackWrapper and then doing nothing with it but reading the property set in the constructor?